File: ToyInterpreter 2/src/Model/ExeStack.java
package Model;

import java.util.ArrayList;
import java.util.Observer;
import java.util.Stack;

import mobserver.CObserver;
import mobserver.StackObserver;
import Model.Statement.Statement;
/*
public class ExeStack implements IStack{
	private final Statement[] st;
	private int size;
	public ExeStack() {
		st=new Statement[Constants.stackSize];
	}
	public void push(Statement s) {
		st[size++] = s;
	}
	
	public Statement top() {
		return st[size-1];
	}
	
	public void pop() {
		--size;
	}
	public boolean isEmpty() {
		return size == 0;
	}
}
*/

public class ExeStack implements IStack,java.io.Serializable {

	private final Stack<Statement> st = new Stack<Statement>();
	private final ArrayList<StackObserver> o = new ArrayList<StackObserver>();
	
	public void addObserver(StackObserver ob) {
		o.add(ob);
	}
	
	private void notifyObservers() {
		for(int i = 0; i < o.size(); ++i) 
			o.get(i).onUpdate(this);
	}
	
	@Override
	public void push(Statement s) {
		st.push(s);
		notifyObservers();
		
	}

	@Override
	public Statement top() {
		return st.peek();
	}

	@Override
	public void pop() {
		st.pop();
	}

	@Override
	public boolean isEmpty() {
		return st.isEmpty();
	}
	
	public String toString() {
		if(isEmpty()) return "Exe stack is empty!\n";
		String r="Exe stack:\n";
		for(Statement s : st) {
			r += "\t" + s.toString() +"\n";
		}
		return r;
	}
}

File: ToyInterpreter/ToyInterpreter/Model/ExeStack.cs
using System;
using ToyInterpreter.Model.MStatement;
using System.Collections.Generic;
using ToyInterpreter.Model.MObserver;

namespace ToyInterpreter.Model
{
	/*public class ExeStack : IStack{
		private readonly Statement[] st;
		private int size;
		public ExeStack() {
			st=new Statement[Constants.stackSize];
		}
		public void push(Statement s) {
			st[size++] = s;
		}

		public Statement top() {
			return st[size-1];
		}

		public void pop() {
			--size;
		}
		public bool isEmpty() {
			return size == 0;
		}
	}*/
	[Serializable()]
	public class ExeStack : IStack{
		private readonly Stack<Statement> st;
		private readonly List<StackObserver> o;

		public ExeStack() {
			st = new Stack<Statement>();
			o = new List<StackObserver> ();
		}

		private void notifyObservers() {
			for(int i = 0; i < o.Count; ++i) o[i].onUpdate(this);
		}

		public void addObserver(StackObserver ob) {
			o.Add(ob);
		}

		public void push(Statement s) {
			st.Push (s);
			notifyObservers ();
		}

		public Statement top() {

			return st.Peek ();
		}

		public void pop() {
			st.Pop ();
		}
		public bool isEmpty() {
			return st.Count == 0;
		}
		public String toString() {
			if(isEmpty()) return "Exe stack is empty!\n";
			String r="Exe stack:\n";
			foreach (Statement s in st) {
				r += "\t" + s.toString() +"\n";
			}
			return r;
		}
	}
}

